src/main/java/build/buildfarm/common/resources/ResourceParser.java

```diff
@@ -0,0 +1,201 @@
+ *     utilities are for parsing "resource name" and extracting the information into easy to work
+ *     with types.
+ */
+public class ResourceParser {
+  /**
+   * @brief Categorize the resource type by analyzing a resource name URI.
+   * @details The URI is parsed to indentify one of the possible Resource types it is in reference
+   *     to.
+   * @param resourceName The resource name URI defined by REAPI.
+   * @return The resource type derived from the URI.
+   * @note Suggested return identifier: resourceType.
+   */
+  public static Resource.TypeCase getResourceType(String resourceName) {
+    // Each "resource name" starts with an optional "instance name" which might consist of multiple
+    // path segments Each "resource name" might also end with "optional metadata" which might also
+    // consist of multiple path segments In order to parse the resource type effectively and without
+    // relying on a segment index, REAPI gives the following guarantee: "To simplify parsing, the
+    // instance_name segments cannot equal any of the following keywords: `blobs`, `uploads`,
+    // `actions`, `actionResults`, `operations`,`capabilities` or `compressed-blobs`." Therefore we
+    // have to scan linearly to identify the resource type by the first known keyword.
+    // https://github.com/bazelbuild/remote-apis/blob/04784f4a830cc0df1f419a492cde9fc323f728db/build/bazel/remote/execution/v2/remote_execution.proto#L215
+    HashMap<String, Resource.TypeCase> keywords = keywordResourceMap();
+
+    // Find the first instance of such a keyword.
+    String type =
+        Stream.of(tokenize(resourceName))
+            .filter(segment -> keywords.containsKey(segment))
+            .findFirst()
+            .orElse("");
+
+    // Return the resource type. If no resource type was identified, return null.
+    return keywords.get(type);
+  }
+  /**
+   * @brief Parse the resource name into a specific type.
+   * @details Assumes the URI is already is already in the upload blob format.
+   * @param resourceName The resource name URI defined by REAPI.
+   * @return The resource type derived from the URI.
+   * @note Suggested return identifier: request.
+   */
+  public static UploadBlobRequest parseUploadBlobRequest(String resourceName) {
+    // Find the index of the keyword.  This will give us an initial index to extract information
+    // from.
+    String[] segments = tokenize(resourceName);
+    MutableInt index = startIndex(segments, Resource.TypeCase.UPLOAD_BLOB_REQUEST);
+
+    // Extract each of the following segments:
+    // `{instance_name}/uploads/{uuid}/blobs/{hash}/{size}{/optional_metadata}` or
+    // `{instance_name}/uploads/{uuid}/compressed-blobs/{compressor}/{uncompressed_hash}/{uncompressed_size}{/optional_metadata}`
+    UploadBlobRequest.Builder builder = UploadBlobRequest.newBuilder();
+    builder.setInstanceName(
+        String.join("/", Arrays.copyOfRange(segments, 0, index.getAndIncrement())));
```

Comment:
please add an appropriate named `static final` separator constant and a `resourcePath` method that does the join over components - fewer magic numbers, fewer convoluted `String.join`s

Reply:
added